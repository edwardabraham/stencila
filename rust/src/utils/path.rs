use lexiclean::Lexiclean;
use std::path::{Path, PathBuf};

/// Merge paths
///
/// Differs from `path.join()` in that it joins the *parent* of `path1`, not `path1` itself.
/// Does not canonicalize the path, or check for its existence.
pub fn merge<P1: AsRef<Path>, P2: AsRef<Path>>(path1: P1, path2: P2) -> PathBuf {
    path1.as_ref().join("..").join(path2).lexiclean()
}

#[cfg(test)]
mod test {
    use super::*;
    use path_slash::PathBufExt;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_merge() {
        assert_eq!(merge("a", "c").to_slash_lossy(), "c");
        assert_eq!(merge("a/b", "c").to_slash_lossy(), "a/c");
        assert_eq!(merge("a/b/../d/e", "c").to_slash_lossy(), "a/d/c")
    }
}
