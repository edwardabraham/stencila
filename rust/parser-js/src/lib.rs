use once_cell::sync::Lazy;
use parser_treesitter::{
    apply_comment_tags,
    eyre::Result,
    formats::Format,
    graph_triples::{relations, resources, Pair, Pairs},
    path_utils,
    utils::remove_quotes,
    Capture, Parser, ParserTrait, TreesitterParser,
};
use std::path::Path;

/// Tree-sitter based parser for JavaScript
static PARSER: Lazy<TreesitterParser> =
    Lazy::new(|| TreesitterParser::new(tree_sitter_javascript::language(), QUERY));

/// Tree-sitter AST query for JavaScript
///
/// Made public for use by `parser-ts`.
pub const QUERY: &str = r#"
(program (comment) @comment)

(import_statement
    source: (string) @module
)
(call_expression
    function: [(import)(identifier)] @function (#match? @function "^import|require$")
    arguments: (arguments . (string) @module)
)

(call_expression
    function: [
        (
            (identifier) @function (#match? @function "^readFile")
        )
        (
            member_expression
                object: (_)
                property: (property_identifier)  @function (#match? @function "^readFile")
        )
    ]
    arguments: (arguments . (string) @path)
)

(call_expression
    function: [
        (
            (identifier) @function (#match? @function "^writeFile")
        )
        (
            member_expression
                object: (_)
                property: (property_identifier)  @function (#match? @function "^writeFile")
        )
    ]
    arguments: (arguments . (string) @path)
)

(program
    [
        (expression_statement
            (assignment_expression
                left: (identifier) @name
                right: (_) @value
            )
        )
        (variable_declaration
            (variable_declarator
                name: (identifier) @name
                .
                value: (_) @value
            )
        )
        (lexical_declaration
            (variable_declarator
                name: (identifier) @name
                .
                value: (_) @value
            )
        )
        (export_statement
            declaration: (lexical_declaration
                (variable_declarator
                    name: (identifier) @name
                    .
                    value: (_) @value
                )
            )
        )
    ]
)
(program
    [
        (function_declaration
            name: (identifier) @name
        )
        (export_statement
            declaration: (function_declaration
                name: (identifier) @name
            )
        )
    ]
)

((identifier) @identifer)
"#;

/// Handle a pattern match
///
/// Made public for use by `parser-ts`.
pub fn handle_patterns(
    path: &Path,
    code: &[u8],
    pattern: &usize,
    captures: &[Capture],
) -> Option<Pair> {
    match pattern {
        1 | 2 => {
            // Imports a module using `import` or `require`
            let capture = match pattern {
                1 => &captures[0],
                2 => &captures[1],
                _ => unreachable!(),
            };
            let range = capture.range;
            let module = remove_quotes(&capture.text.clone());
            let object = if module.starts_with("./") {
                resources::file(&path_utils::merge(path, &[&module, ".js"].concat()))
            } else {
                resources::module("javascript", &module)
            };
            Some((relations::uses(range), object))
        }
        3 => {
            // Reads a file
            Some((
                relations::reads(captures[1].range),
                resources::file(&path_utils::merge(path, remove_quotes(&captures[1].text))),
            ))
        }
        4 => {
            // Writes a file
            Some((
                relations::writes(captures[1].range),
                resources::file(&path_utils::merge(path, remove_quotes(&captures[1].text))),
            ))
        }
        5 | 6 => {
            // Assigns a symbol at the top level of the module
            let range = captures[0].range;
            let name = captures[0].text.clone();
            let kind = match pattern {
                5 => match captures[1].node.kind() {
                    "true" | "false" => "Boolean",
                    "number" => "Number",
                    "string" => "String",
                    "array" => "Array",
                    "object" => "Object",
                    "arrow_function" => "Function",
                    _ => "",
                },
                6 => "Function",
                _ => unreachable!(),
            };
            Some((
                relations::assigns(range),
                resources::symbol(path, &name, kind),
            ))
        }
        7 => {
            // Uses an identifier assigned elsewhere
            let node = captures[0].node;
            let range = captures[0].range;
            let symbol = captures[0].text.clone();

            let mut parent = node.parent();
            while let Some(parent_node) = parent {
                match parent_node.kind() {
                    // Skip identifiers in import statements
                    // Could just skip children of `import_statement`, but specifying others in tree
                    // results in an earlier return while walking up tree.
                    "import_statement" | "import_clause" | "named_imports" | "import_specifier" => {
                        return None
                    }
                    // Skip identifiers that are the `name` of a declaration
                    "variable_declarator" => {
                        if Some(node) == parent_node.child_by_field_name("name") {
                            return None;
                        }
                    }
                    // Skip identifiers that are the `left` of an assignment
                    "assignment_expression" => {
                        if Some(node) == parent_node.child_by_field_name("left") {
                            return None;
                        }
                    }
                    // Skip any identifier used in a function
                    "function_declaration" | "arrow_function" | "formal_parameters" => {
                        return None;
                    }
                    // Skip identifiers that are the `left` of a for in loop, or that refer to it
                    // within the loop
                    "for_in_statement" => {
                        if let Some(left) = parent_node.child_by_field_name("left") {
                            if left == node || left.utf8_text(code).unwrap() == symbol {
                                return None;
                            }
                        }
                    }
                    // Skip identifiers that are the `name` of the variable in a for loop, or
                    // that refer to it within the loop
                    "for_statement" => {
                        if let Some(name) = parent_node
                            .child_by_field_name("initializer")
                            .and_then(|node| node.named_child(0)) // variable_declarator
                            .and_then(|node| node.child_by_field_name("name"))
                        {
                            if name == node || name.utf8_text(code).unwrap() == symbol {
                                return None;
                            }
                        }
                    }
                    _ => {}
                }
                parent = parent_node.parent();
            }

            Some((relations::uses(range), resources::symbol(path, &symbol, "")))
        }
        _ => None,
    }
}

/// A parser for JavaScript
pub struct JsParser {}

impl ParserTrait for JsParser {
    fn spec() -> Parser {
        Parser {
            language: Format::JavaScript.spec().title,
        }
    }

    fn parse(path: &Path, code: &str) -> Result<Pairs> {
        let code = code.as_bytes();
        let tree = PARSER.parse(code);
        let matches = PARSER.query(code, &tree);

        let relations = matches
            .iter()
            .filter_map(|(pattern, capture)| handle_patterns(path, code, pattern, capture))
            .collect();

        let pairs = apply_comment_tags(path, &Self::spec().language, matches, 0, relations);
        Ok(pairs)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use test_snaps::{insta::assert_json_snapshot, snapshot_fixtures};
    use test_utils::fixtures;

    #[test]
    fn parse_js_fragments() {
        snapshot_fixtures("fragments/js/*.js", |path| {
            let code = std::fs::read_to_string(path).expect("Unable to read");
            let path = path.strip_prefix(fixtures()).expect("Unable to strip");
            let pairs = JsParser::parse(path, &code).expect("Unable to parse");
            assert_json_snapshot!(pairs);
        })
    }
}
