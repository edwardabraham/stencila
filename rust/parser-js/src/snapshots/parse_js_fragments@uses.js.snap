---
source: rust/parser-js/src/lib.rs
assertion_line: 284
expression: resource_info
input_file: fixtures/fragments/js/uses.js

---
{
  "resource": {
    "type": "Code",
    "path": "fragments/js/uses.js",
    "id": "",
    "kind": "SoftwareSourceCode",
    "language": "JavaScript"
  },
  "relations": [
    [
      {
        "type": "Use",
        "range": [
          0,
          0,
          0,
          4
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var1",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          1,
          0,
          1,
          4
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var2",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          1,
          7,
          1,
          11
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var3",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          2,
          0,
          2,
          4
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "fun1",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          3,
          0,
          3,
          4
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "fun2",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          3,
          5,
          3,
          9
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var4",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          3,
          11,
          3,
          15
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var5",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          4,
          0,
          4,
          4
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var6",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          4,
          5,
          4,
          9
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var7",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          6,
          4,
          6,
          8
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var8",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          6,
          11,
          6,
          15
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var9",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          7,
          13,
          7,
          18
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var10",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          12,
          19,
          12,
          23
        ]
      },
      {
        "type": "Module",
        "language": "javascript",
        "name": "fs"
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          13,
          21,
          13,
          25
        ]
      },
      {
        "type": "Module",
        "language": "javascript",
        "name": "fs"
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          17,
          15,
          17,
          20
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var11",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          18,
          4,
          18,
          9
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var12",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          21,
          26,
          21,
          31
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var13",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          27,
          0,
          27,
          5
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var14",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          28,
          0,
          28,
          5
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var14",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          29,
          0,
          29,
          5
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var14",
        "kind": ""
      }
    ],
    [
      {
        "type": "Use",
        "range": [
          29,
          13,
          29,
          18
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "var15",
        "kind": ""
      }
    ],
    [
      {
        "type": "Assign",
        "range": [
          34,
          9,
          34,
          14
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "func1",
        "kind": "Function"
      }
    ],
    [
      {
        "type": "Assign",
        "range": [
          38,
          6,
          38,
          11
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "func2",
        "kind": "Function"
      }
    ],
    [
      {
        "type": "Assign",
        "range": [
          42,
          0,
          42,
          6
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "assign",
        "kind": "Number"
      }
    ],
    [
      {
        "type": "Assign",
        "range": [
          43,
          4,
          43,
          10
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "assign",
        "kind": "Number"
      }
    ],
    [
      {
        "type": "Assign",
        "range": [
          44,
          4,
          44,
          10
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "assign",
        "kind": "Number"
      }
    ],
    [
      {
        "type": "Assign",
        "range": [
          45,
          6,
          45,
          12
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/uses.js",
        "name": "assign",
        "kind": "Number"
      }
    ]
  ],
  "autorun": "Needed",
  "compile_digest": "G6WG8Hz_jBsK94XcrK0hNXQLkTHkPP0F0cbBLL4iMfA.bIguRn2OwB43TaEj9RIM9CTEiCdK3tIjmaLLDAHUZ9o..0"
}
